src/types/dict.rs

```diff
@@ -372,85 +396,333 @@ fn dict_len(dict: &Bound<'_, PyDict>) -> Py_ssize_t {
+                move |(), x| match f(x) {
+                    Some(x) => ControlFlow::Break(x),
+                    None => ControlFlow::Continue(()),
+                }
+            }
+
+            match iter.try_fold((), check(f)) {
+                ControlFlow::Continue(_) => None,
+                ControlFlow::Break(x) => Some(x),
+            }
+        })
+    }
+
+    #[inline]
+    #[cfg(all(Py_GIL_DISABLED, not(feature = "nightly")))]
+    fn position<P>(&mut self, predicate: P) -> Option<usize>
+    where
+        Self: Sized,
+        P: FnMut(Self::Item) -> bool,
+    {
+        self.with_critical_section(|iter| {
+            #[inline]
+            fn check<'a, T>(
+                mut predicate: impl FnMut(T) -> bool + 'a,
+                acc: &'a mut usize,
+            ) -> impl FnMut((), T) -> ControlFlow<usize, ()> + 'a {
+                move |_, x| {
+                    if predicate(x) {
+                        ControlFlow::Break(*acc)
+                    } else {
+                        *acc += 1;
+                        ControlFlow::Continue(())
+                    }
+                }
+            }
+
+            let mut acc = 0;
+            match iter.try_fold((), check(predicate, &mut acc)) {
+                ControlFlow::Continue(_) => None,
+                ControlFlow::Break(x) => Some(x),
+            }
+        })
+    }
+}
+
+#[inline]
+#[cfg(Py_GIL_DISABLED)]
+pub fn with_critical_section<F, R>(object: &Bound<'_, PyAny>, f: F) -> R
+where
+    F: FnOnce() -> R,
+{
+    struct Guard(ffi::PyCriticalSection);
+
+    impl Drop for Guard {
+        fn drop(&mut self) {
+            unsafe {
+                ffi::PyCriticalSection_End(&mut self.0);
+            }
+        }
+    }
+
+    let _guard = unsafe {
+        let mut section = std::mem::zeroed();
+        ffi::PyCriticalSection_Begin(&mut section, object.as_ptr());
+        Guard(section)
+    };
+    f()
 }
 
 impl<'py> ExactSizeIterator for BoundDictIterator<'py> {
     fn len(&self) -> usize {
-        self.len as usize
+        match self {
+            BoundDictIterator::ItemIter { remaining, .. } => *remaining as usize,
+            BoundDictIterator::DictIter { remaining, .. } => *remaining as usize,
+        }
     }
 }
 
 impl<'py> BoundDictIterator<'py> {
     fn new(dict: Bound<'py, PyDict>) -> Self {
-        let len = dict_len(&dict);
-        BoundDictIterator {
-            dict,
-            ppos: 0,
-            di_used: len,
-            len,
+        let remaining = dict_len(&dict);
+
+        if dict.is_exact_instance_of::<PyDict>() {
+            return BoundDictIterator::DictIter {
+                dict,
+                ppos: 0,
+                di_used: remaining,
+                remaining,
+            };
+        };
+
+        let items = dict.call_method0(intern!(dict.py(), "items")).unwrap();
+        let iter = PyIterator::from_object(&items).unwrap();
+        BoundDictIterator::ItemIter { iter, remaining }
```

Comment:
I think there's a possible design decision to be had about the subclasses and `.items()` call. Do we definitely want to do this? We essentially already do this for sets, so there is a good argument we should do this here for consistency. (And I guess similar for lists?) Maybe this wants to be split into its own PR / discussion.

See also #4490 which is very related.

Reply:
If removing the `ItemIter` from this MR makes it easier to review then I'm happy to open a second one and remove the changes from this MR.

Reply:
I think I would still like it split out, yes please. I think it's _probably_ correct in the long run down that road, but I also fear that (a) no other dict methods currently respect subclasses and (b) we probably want to synchronise such a change with lists and tuple iterators too, as well as consider what other methods to adjust at the same time. (Probably in combination with a decision on #4490.)